framework/src/tinyalsa/tinyalsa.c

```diff
@@ -497,53 +497,62 @@ int pcm_writei(struct pcm *pcm, const void *data, unsigned int frame_count)
 	}
 
 	pending = pcm_frames_to_bytes(pcm, frame_count);
-
 	while (pending > 0) {
-		nbytes = pending > pcm_frames_to_bytes(pcm, pcm->buffer_size) ? pcm_frames_to_bytes(pcm, pcm->buffer_size) : pending;
-		if (pcm->buf_idx < pcm->buffer_cnt) {
-			/* If we have empty buffers, fill them first */
-			memcpy(pcm->pBuffers[pcm->buf_idx]->samp, (char *)data + offset, nbytes);
-			apb = pcm->pBuffers[pcm->buf_idx];
-			pcm->buf_idx++;
+		if (pcm->next_buf != NULL) {
+			apb = pcm->next_buf;
 		} else {
-			/* We dont have any empty buffers. wait for deque message from kernel */
-			size = mq_receive(pcm->mq, (FAR char *)&msg, sizeof(msg), &prio);
-			if (size != sizeof(msg)) {
-				/* Interrupted by a signal? What to do? */
-				return oops(pcm, EINTR, "Interrupted while waiting for deque message from kernel\n");
-			}
-			if (msg.msgId == AUDIO_MSG_DEQUEUE) {
-				apb = (struct ap_buffer_s *)msg.u.pPtr;
-				memcpy(apb->samp, (char *)data + offset, nbytes);
-			} else if (msg.msgId == AUDIO_MSG_XRUN) {
-				/* Underrun to be handled by client */
-				return -EPIPE;
+			/* First time we will use empty pBuffer, and then wait until one of them dequeued */
+			if (pcm->buf_idx < pcm->buffer_cnt) {
+				apb = pcm->pBuffers[pcm->buf_idx];
+				pcm->buf_idx++;
 			} else {
-				return oops(pcm, EINTR, "Recieved unexpected msg (id = %d) while waiting for deque message from kernel\n", msg.msgId);
+				/* We dont have any empty buffers. wait for deque message from kernel */
+				size = mq_receive(pcm->mq, (FAR char *)&msg, sizeof(msg), &prio);
+				if (size != sizeof(msg)) {
+					/* Interrupted by a signal? What to do? */
+					return oops(pcm, EINTR, "Interrupted while waiting for deque message from kernel\n");
+				}
+				if (msg.msgId == AUDIO_MSG_DEQUEUE) {
+					apb = (struct ap_buffer_s *)msg.u.pPtr;
+				} else if (msg.msgId == AUDIO_MSG_XRUN) {
+					/* Underrun to be handled by client */
+					return -EPIPE;
+				} else {
+					return oops(pcm, EINTR, "Recieved unexpected msg (id = %d) while waiting for deque message from kernel\n", msg.msgId);
+				}
 			}
 		}
-
 		/* Buffer is ready. Enque it */
 #ifdef CONFIG_AUDIO_MULTI_SESSION
 		bufdesc.session = pcm->session;
 #endif
-		apb->nbytes = nbytes;
+		nbytes = pending > pcm->buffer_size - apb->nbytes ? pcm->buffer_size - apb->nbytes : pending;	
+		memcpy(apb->samp + apb->nbytes, (char *)data + offset, nbytes);
+		medvdbg("pending : %d nbytes : %d offset : %d apb : %x\n", pending, nbytes, offset, apb);
+		apb->nbytes += nbytes;
 		apb->curbyte = 0;
 		apb->flags = 0;
```

Comment:
These two look repeated, may be we can assign them when the apb is assigend for the first time? You can ignore this, not major.